### PR TITLE
Closes #1059: Async processing of Event -> Action mapping

### LINKED ISSUE #1059: Adding actions causes Postgres out of memory with partitions
**Describe the bug**

Report from OpenRent that adding a simple $autocapture action causes Postgres to run out of memory. I think this has something to do with the fact that they've got things partitioned.

**To Reproduce**
1. TBD
2. 
3. 

**Expected behavior**

It should work.

**Screenshots**

**Hosted or self hosted?:**
- self-hosted

**Additional context**

### PR DESCRIPTION
**Pull request by fuziontech**
Closes #1059 
- Add Env Var `ASYNC_EVENT_ACTION_MAPPING` for deactivating inline Event Action lookups to prevent gnarly query generation resulting in Postgres OOM
- Setup a Celery Beat task to do Action -> Event mapping every minute on all Actions (we may want to tune this or set this as the env var parameter)

Already test coverage on `Action.calculate_events()` in a few places
https://github.com/PostHog/posthog/blob/da5a48c33d059be9d5c0cb4259524560454e2df8/posthog/api/test/test_action.py#L84

### CODE REVIEW

**Comment by fuziontech** [comment]
From my testing this brings the mean time to process an event from 140 ms to 40 ms

**Review by mariusandra — COMMENTED** [review]
Basically this works for me, but I don't know if it solves the issue. What will happen if the `calculate_events` function takes more than a minute. I'd bet we'll get some really weird deadlocks ☠️🔒?

Looking at `calculate_events`, it seems that we are deleting and recreating all event_actions every time we match events to actions. I'm old enough in this cpy to remember when that got added 😆 . This was done to offload all the processing to the DB and away from Django. 

image

However, I fear this calculation is causing us serious problems now.

When I look at the IDs in this table:

image

... and hit reload 60sec later after `calculate_all_actions` runs, it's clear that there has been a lot of reshuffling in the database, even though no new events came in... and no *new* events were matched to any actions. Just all the old ones got deleted and then recreated.

Here are some more screenshots from the DB:

image
image
image

The diff in numbers:

image

I'm afraid I'll run out of 64 bit integers soon if this keeps happening.

And every time this happens, postgres calls up a bunch of friends and goes on a rampage:

image

There must be some way to only create new matched pairs... or keep track of "events processed until X o'clock for this action" and only process new ones... or something like that. I don't know :).

So while this PR is technically solid, I'm not sure if it'll have the desired end effect.

**Comment by mariusandra** [comment]
With actions I think a large part of the complexity comes from the need to recalculate all action and event pairs when some property/filter changes.

What if we remove that requirement and make all actions immutable? You can only change the name, but if you want to change the filters, we'll silently create a new action for you instead. 

Could this bring down some of the complexity?

File: posthog/celery.py [hunk 1]
```diff
@@ -32,6 +33,9 @@
 redis_instance = redis.from_url(settings.REDIS_URL, db=0)
 
 
+ACTION_EVENT_MAPPING_INTERVAL_MINUTES = 0.5
```
Comment (fuziontech) [inline posthog/celery.py:36]: This is just for debugging, will need to increase to 5+ minutes

**Comment by fuziontech** [comment]
The running out of 64bit integer's for the PK/ID is a super valid point. I hadn't thought of that. Something I would love to eventually get us over to is using UUID's for all PK's.

I did change this up to be incremental, only deleting and calculating relationships since the last time the action was recalculated. Should reduce that churn significantly.

**Review by mariusandra — APPROVED** [review]
The first time I ran it, postgres went wild as expected to calculate the first batch:



Since then it was quiet and maximally using one core to do its work.

30sec was not enough for 1M events. 5min surely would be, yet we have users with even more events than that. It might useful for the users who will switch this on to run the first calculation manually before enabling the flag.

Recalculating the actions however takes 19 seconds for me every round, which seems slower than expected for a "select nothing and do nothing with it" type query. Might be worth adding an index for `created_at` 🤓 



(The output shown is for `print(action.last_calculated)` for each action. )

For some actions the `last_calculated` dates don't change after the first calculation. I guess they're the ones that returned `EmptyResultSet`. Would it be worth setting `last_calculated` for them as well... to possibly further limit some future index lookups? Otherwise data may start spanning partitions or some other weird case that will bite us one day?

In any case, some nits below, but otherwise approved from my side!

**Comment by mariusandra** [comment]
Ok, it's definitely worth running the first calculation manually. For me I now have each action 4x in the database since it took at least 4*30sec to calculate them for me. 

1M events in 5min is again fine, but what about 100M? We don't want duplicates at that scale...

image

**Comment by mariusandra** [comment]
One idea could be to set "last_calculated_at" to right now if it's nil... and then run another job to back-calculate for the past. Another idea is to prevent recalculating if `self.is_calculating = True`... but that brings up the question of crashed jobs 🤔

**Comment by timgl** [comment]
@fuziontech is this ready for review?

File: posthog/celery.py [hunk 2]
```diff
@@ -42,6 +43,15 @@ def setup_periodic_tasks(sender, **kwargs):
     sender.add_periodic_task(15 * 60, calculate_cohort.s(), name="debug")
     sender.add_periodic_task(600, check_cached_items.s(), name="check dashboard items")
 
+    action_event_mapping_interval_minutes = settings.ACTION_EVENT_MAPPING_INTERVAL_MINUTES
```
Comment (timgl) [inline posthog/celery.py:46]: I don't think this should be an env variable. Let's just pick a sensible default. Every 10 minutes seems fine to me

File: posthog/models/event.py [hunk 3]
```diff
@@ -354,6 +369,7 @@ def actions(self) -> List:
         filtered_actions = [action for action in actions if getattr(event, "action_{}".format(action.pk))]
         return filtered_actions
 
+    created_at: models.DateTimeField = models.DateTimeField(auto_now_add=True, blank=True)
```
Comment (timgl) [inline posthog/models/event.py:372]: this should have `, null=True`

**Pull request merged**